File: frontend/Invoice Frontend/src/Context/GlobalStateProvider.js
import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios';
import baseUrl from '../api/utils';
import favicon from "../images/favicon.ico"

// Create the context
export const GlobalStateContext = createContext();

const token =  sessionStorage.getItem("token")
const role=sessionStorage.getItem("role")
// Create a provider component
export const GlobalStateProvider = ({ children }) => {
  const [displayname, setDisplayname] = useState({
              admin_name: "",
              trainer_name: "",
              student_name: "", 
  });
  const [siteSettings, setsiteSettings] = useState({
    siteUrl: "",
    title:"",
    sitelogo: null,
    siteicon: null,
    titleicon: null,
  });
  const[Activeprofile,setActiveProfile]=useState("")

  const [generalSettings, setGeneralSettings] = useState({
  gstinNumber: true,
  estimateQuotation: true,
  salesInvoiceOrder: true,
  amountDecimalPlaces: 2,
  otpservice:true,
});

// Add state
const [gstSettings, setGstSettings] = useState({
  enableGST: true,
  enableHSN: true,
  additionalCess: false,
  enablePlaceOfSupply: true,
});

const [itemSettings, setItemSettings] = useState({
  enableItem: true,
  stockMaintenance: true,
  showLowStockDialog: true,
  itemsUnit: true,
  defaultUnit: "",
  itemCategory: true,
  description: false,
  itemWiseTax: true,
  itemWiseDiscount: true,
  quantityDecimalPlaces: 2,
  wholesalePrice: true,
  mrp: false,
  calculateTaxBasedOnMrp: false,
  expDate: true,
  expDateFormat: "mm/yy",
  mfgDate: true,
  mfgDateFormat: "dd/mm/yy",
  modelNo: false,
  size: false,
});


const [lowStockItems, setLowStockItems] = useState([]);



  useEffect(() => {
    const fetchactiveprofile=async()=>{
      try{
        const active = await axios.get(`${baseUrl}/Active/Environment`);
        if (active?.data?.environment) {
          sessionStorage.setItem("Activeprofile", active.data.environment);
          setActiveProfile(active.data.environment);
        } 
        if (active?.data?.currency) {
          sessionStorage.setItem("Currency", active.data.currency);
        } 
      }catch(error){
        console.log(error)
      }
    }
    fetchactiveprofile();
    
     // 👇 Add this new function
      const fetchGeneralSettings = async () => {
        if (token) {
          try {
            const res = await axios.get(`${baseUrl}/settings/general/get`, {
              headers: { Authorization: token }
            });
            setGeneralSettings(res.data);
          } catch (err) {
            console.error("Failed to load general settings", err);
          }
        }
      };
    fetchGeneralSettings();


    // Add gst fetch inside useEffect
    const fetchGstSettings = async () => {
      if (token) {
        try {
          const res = await axios.get(`${baseUrl}/settings/gst/get`, {
            headers: { Authorization: token }
          });
          setGstSettings(res.data);
        } catch (err) {
          console.error("Failed to load GST settings", err);
        }
      }
    };

    fetchGstSettings();

// 👇 Add this new function
const fetchItemSettings = async () => {
  if (token) {
    try {
      const res = await axios.get(`${baseUrl}/settings/item/get`, {
        headers: { Authorization: token }
      });

     setItemSettings(res.data);
    } catch (err) {
      console.error("Failed to load item settings", err);
    }
  }
};

fetchItemSettings();


    const fetchLabels = async () => {
      try {
        // Check if data is already in sessionStorage
        const cachedSettings = sessionStorage.getItem("siteSettings");
        if (cachedSettings) {
          const parsedSettings = JSON.parse(cachedSettings);
          setsiteSettings(parsedSettings);
          updateSiteMeta(parsedSettings); 
        } else {
          let response;
          if (token && role==="ADMIN") {
           
            response = await axios.get(`${baseUrl}/Get/labellings`, {
              headers: {
                Authorization: token,
              },
            });
          
          } else {
            response = await axios.get(`${baseUrl}/all/get/labellings`);
          }

          setsiteSettings(response.data);
          sessionStorage.setItem("siteSettings", JSON.stringify(response.data)); // Cache data
          updateSiteMeta(response.data);
        }
      } catch (error) {
        console.error("Error fetching site settings:", error);
      }
    };

    const updateSiteMeta = (settings) => {
      // Update document title
      document.title = settings.title || "Invoice Bill";

      // Update favicon
      const faviconLink = document.querySelector("link[rel='icon']");
      const faviconHref = settings.titleicon
        ? `data:image/x-icon;base64,${settings.titleicon}`
        : favicon; // Default to imported favicon.ico

      if (faviconLink) {
        faviconLink.href = faviconHref;
      } else {
        const newLink = document.createElement("link");
        newLink.rel = "icon";
        newLink.href = faviconHref;
        document.head.appendChild(newLink);
      }
    };

//     const fetchLowStockItems = async () => {
//   const token = sessionStorage.getItem("token"); 
//   if (token) {
//     try {
//       const res = await axios.get(`${baseUrl}/getLowStockItems`, {
//         headers: { Authorization: token }
//       });
//       setLowStockItems(res.data);
//     } catch (err) {
//       console.error("Failed to load low stock items", err);
//     }
//   }
// };
// fetchLowStockItems();

    fetchLabels();
  }, []);

  

  useEffect(() => {
    const fetchDisplayNameSettings = async () => {
      const savedDisplayName = sessionStorage.getItem("displayname");

      if (savedDisplayName) {
        // If data is found in sessionStorage, use it
        setDisplayname(JSON.parse(savedDisplayName));
      } else if (token) {
        try {
          const response = await axios.get(`${baseUrl}/get/displayName`, {
            headers: {
              Authorization: token,
            },
          });
          if (response.status === 200) {
            const data = response.data;
            setDisplayname(data);
            sessionStorage.setItem("displayname", JSON.stringify(data)); // Save in localStorage
        } else if (response.status === 204) {
          setDisplayname([]);
          sessionStorage.setItem("displayname", JSON.stringify([]));
        }   
          // else if(response.status === 204){
            
          //     // const data = response.data;
          //     // setDisplayname("");
          //     // sessionStorage.setItem("displayname", JSON.stringify(data)); // Save in localStorage
            
          // }
        } catch (error) {
          if (error.response.status === 404) {
            const emptyData = {
              admin_name: "",
              trainer_name: "",
              student_name: "",
            };
            setDisplayname(emptyData);
            sessionStorage.setItem("displayname", JSON.stringify(emptyData));
          }else{
            throw error
          }
        }
      }
    };

    fetchDisplayNameSettings();
  }, [token]);

  return (
    <GlobalStateContext.Provider value={{ displayname ,siteSettings,Activeprofile,generalSettings,setGeneralSettings,
      gstSettings, setGstSettings ,itemSettings, setItemSettings, lowStockItems,setLowStockItems
    }}>
      {children}
    </GlobalStateContext.Provider>
  );
};

File: invoice_flutter_final/lib/core/state/global_state_provider.dart
import 'package:flutter/foundation.dart';
import '../api/api_client.dart';
import '../session/session_manager.dart';

/// Flutter equivalent of `src/Context/GlobalStateProvider.js`.
///
/// The original React provider, on mount, fired 4 requests:
///   GET /Active/Environment        -> Activeprofile, Currency
///   GET /settings/general/get      -> generalSettings   (auth'd)
///   GET /settings/gst/get          -> gstSettings        (auth'd)
///   GET /settings/item/get         -> itemSettings       (auth'd)
///   GET /Get/labellings            -> siteSettings       (auth'd, ADMIN only)
///   GET /all/get/labellings        -> siteSettings       (everyone else)
///   GET /get/displayName           -> displayname        (auth'd)
///
/// Same endpoints, same fallbacks/defaults, same caching-in-storage
/// behavior (siteSettings + displayname were cached in sessionStorage in
/// React; mirrored here via SessionManager/SharedPreferences).
class GlobalStateProvider extends ChangeNotifier {
  Map<String, dynamic> displayname = const {
    'admin_name': '',
    'trainer_name': '',
    'student_name': '',
  };

  Map<String, dynamic> siteSettings = const {
    'siteUrl': '',
    'title': '',
    'sitelogo': null,
    'siteicon': null,
    'titleicon': null,
  };

  String activeProfile = '';

  Map<String, dynamic> generalSettings = const {
    'gstinNumber': true,
    'estimateQuotation': true,
    'salesInvoiceOrder': true,
    'amountDecimalPlaces': 2,
    'otpservice': true,
  };

  Map<String, dynamic> gstSettings = const {
    'enableGST': true,
    'enableHSN': true,
    'additionalCess': false,
    'enablePlaceOfSupply': true,
  };

  Map<String, dynamic> itemSettings = const {
    'enableItem': true,
    'stockMaintenance': true,
    'showLowStockDialog': true,
    'itemsUnit': true,
    'defaultUnit': '',
    'itemCategory': true,
    'description': false,
    'itemWiseTax': true,
    'itemWiseDiscount': true,
    'quantityDecimalPlaces': 2,
    'wholesalePrice': true,
    'mrp': false,
    'calculateTaxBasedOnMrp': false,
    'expDate': true,
    'expDateFormat': 'mm/yy',
    'mfgDate': true,
    'mfgDateFormat': 'dd/mm/yy',
    'modelNo': false,
    'size': false,
  };

  List<dynamic> lowStockItems = [];

  /// Mirrors the `setLowStockItems` setter exposed via GlobalStateContext
  /// and called from Header.js's `fetchLowStock`.
  void setLowStockItems(List<dynamic> items) {
    lowStockItems = items;
    notifyListeners();
  }

  /// Mirrors `setGeneralSettings` from GlobalStateContext, called by
  /// `UserSettings/SettingsComponent.js` right after a successful
  /// `POST /settings/general/save`.
  void setGeneralSettings(Map<String, dynamic> settings) {
    generalSettings = {...generalSettings, ...settings};
    notifyListeners();
  }

  /// Mirrors `setGstSettings`, called by `UserSettings/TaxesGST.js` after
  /// `POST /settings/gst/save`.
  void setGstSettings(Map<String, dynamic> settings) {
    gstSettings = {...gstSettings, ...settings};
    notifyListeners();
  }

  /// Mirrors `setItemSettings`, called by `UserSettings/Itemsettings.js`
  /// after `POST /settings/item/save`.
  void setGlobalItemSettings(Map<String, dynamic> settings) {
    itemSettings = {...itemSettings, ...settings};
    notifyListeners();
  }

  /// Mirrors the `sessionStorage.setItem("displayname", ...)` +
  /// GlobalStateContext refresh done by `UserSettings/DisplayName.js`
  /// after a successful save, so the sidebar's custom Trainer/Student
  /// labels update immediately without a full re-login.
  Future<void> setDisplayNameMap(Map<String, dynamic> data) async {
    displayname = data;
    await SessionManager.instance.setDisplayName(data);
    notifyListeners();
  }

  /// Mirrors `Sitesettings.js` writing straight to the labellings cache
  /// after a successful `POST /save/labellings`.
  Future<void> setSiteSettingsMap(Map<String, dynamic> data) async {
    siteSettings = data;
    await SessionManager.instance.setSiteSettings(data);
    notifyListeners();
  }

  Future<void> init() async {
    await Future.wait([
      _fetchActiveProfile(),
      _fetchGeneralSettings(),
      _fetchGstSettings(),
      _fetchItemSettings(),
      _fetchLabels(),
      _fetchDisplayName(),
    ]);
  }

  Future<void> _fetchActiveProfile() async {
    try {
      final res = await ApiClient.instance.dio.get('/Active/Environment');
      final env = res.data?['environment'];
      final cur = res.data?['currency'];
      if (env != null) {
        activeProfile = env.toString();
        await SessionManager.instance.setActiveProfile(activeProfile);
      }
      if (cur != null) {
        await SessionManager.instance.setCurrency(cur.toString());
      }
    } catch (e) {
      debugPrint('Active/Environment error: $e');
    }
    notifyListeners();
  }

  Future<void> _fetchGeneralSettings() async {
    final token = await SessionManager.instance.token;
    if (token == null) return;
    try {
      final res = await ApiClient.instance.dio.get('/settings/general/get');
      // BUGFIX: merge onto defaults instead of replacing the whole map -
      // if the backend response omits a key, it must fall back to the
      // default rather than silently becoming null/missing.
      if (res.data is Map) generalSettings = {...generalSettings, ...Map<String, dynamic>.from(res.data as Map)};
    } catch (e) {
      debugPrint('Failed to load general settings: $e');
    }
    notifyListeners();
  }

  Future<void> _fetchGstSettings() async {
    final token = await SessionManager.instance.token;
    if (token == null) return;
    try {
      final res = await ApiClient.instance.dio.get('/settings/gst/get');
      if (res.data is Map) gstSettings = {...gstSettings, ...Map<String, dynamic>.from(res.data as Map)};
    } catch (e) {
      debugPrint('Failed to load GST settings: $e');
    }
    notifyListeners();
  }

  Future<void> _fetchItemSettings() async {
    final token = await SessionManager.instance.token;
    if (token == null) return;
    try {
      final res = await ApiClient.instance.dio.get('/settings/item/get');
      // BUGFIX (root cause of "low stock notifications never appear"):
      // this used to replace `itemSettings` outright with whatever the
      // backend returned. If that response didn't include
      // `showLowStockDialog` (e.g. an item-settings row saved before that
      // field existed, or the backend simply omits false/never-set
      // fields), the key became missing entirely, `itemSettings
      // ['showLowStockDialog'] == true` evaluated to false, and the low-
      // stock badge/notification was permanently hidden regardless of
      // actual stock levels. Merging onto the default (`true`) means a
      // missing key can no longer silently turn the feature off.
      if (res.data is Map) itemSettings = {...itemSettings, ...Map<String, dynamic>.from(res.data as Map)};
    } catch (e) {
      debugPrint('Failed to load item settings: $e');
    }
    notifyListeners();
  }

  Future<void> _fetchLabels() async {
    try {
      final cached = await SessionManager.instance.siteSettings;
      if (cached != null) {
        siteSettings = cached;
        notifyListeners();
        return;
      }
      final token = await SessionManager.instance.token;
      final role = await SessionManager.instance.role;
      final res = (token != null && role == 'ADMIN')
          ? await ApiClient.instance.dio.get('/Get/labellings')
          : await ApiClient.instance.dio.get('/all/get/labellings');
      // BUGFIX: backend can return "" (empty body) instead of a JSON
      // object when no labels are configured yet - guard the cast so
      // that just falls back to defaults instead of throwing.
      if (res.data is Map) {
        siteSettings = Map<String, dynamic>.from(res.data as Map);
        await SessionManager.instance.setSiteSettings(siteSettings);
      }
    } catch (e) {
      debugPrint('Error fetching site settings: $e');
    }
    notifyListeners();
  }

  Future<void> _fetchDisplayName() async {
    final cached = await SessionManager.instance.displayName;
    if (cached != null) {
      displayname = cached;
      notifyListeners();
      return;
    }
    final token = await SessionManager.instance.token;
    if (token == null) return;
    try {
      final res = await ApiClient.instance.dio.get('/get/displayName');
      if (res.statusCode == 200 && res.data is Map) {
        displayname = Map<String, dynamic>.from(res.data as Map);
        await SessionManager.instance.setDisplayName(displayname);
      } else if (res.statusCode == 204) {
        displayname = {};
        await SessionManager.instance.setDisplayName({});
      }
    } catch (e) {
      // React: on 404 it reset to blank defaults instead of surfacing error.
      displayname = {
        'admin_name': '',
        'trainer_name': '',
        'student_name': '',
      };
      await SessionManager.instance.setDisplayName(displayname);
    }
    notifyListeners();
  }
}
